lib/spack/spack/architecture.py

```diff
@@ -56,49 +56,197 @@
+            self.module_name == other.module_name
+
+    def __ne__(self, other):
+        # This method is necessary as long as we support Python 2. In Python 3
+        # __ne__ defaults to the implementation below
+        return not self == other
+
+    @_ensure_other_is_target
+    def __lt__(self, other):
+        # This is needed to sort deterministically specs in a list.
+        # It doesn't implement a total ordering semantic.
+        return self.micro_architecture.name < other.micro_architecture.name
+
+    def __hash__(self):
+        return hash((self.name, self.module_name))
+
+    @staticmethod
+    def from_dict_or_value(dict_or_value):
+        # A string here represents a generic target (like x86_64 or ppc64) or
+        # a custom micro-architecture
+        if isinstance(dict_or_value, six.string_types):
+            return Target(dict_or_value)
+
+        # TODO: From a dict we actually retrieve much more information than
+        # TODO: just the name. We can use that information to reconstruct an
+        # TODO: "old" micro-architecture or check the current definition.
+        target_info = dict_or_value
+        return Target(target_info['name'])
+
+    def to_dict_or_value(self):
+        """Returns a dict or a value representing the current target.
+
+        String values are used to keep backward compatibility with generic
+        targets, like e.g. x86_64 or ppc64. More specific micro-architectures
+        will return a dictionary which contains information on the name,
+        features, vendor, generation and parents of the current target.
+        """
+        # Generic targets represent either an architecture
+        # family (like x86_64) or a custom micro-architecture
+        if self.micro_architecture.vendor == 'generic':
+            return str(self)
+
+        return syaml_dict([
+            ('name', str(self.micro_architecture.name)),
+            ('vendor', str(self.micro_architecture.vendor)),
+            ('features', sorted(
+                str(x) for x in self.micro_architecture.features
+            )),
+            ('generation', self.micro_architecture.generation),
+            ('parents', [str(x) for x in self.micro_architecture.parents])
+        ])
+
+    def isa_target_for(self, compiler):
+        """Returns the name to be used to optimize for the current target
+        using the compiler passed as argument, or None if the information
+        is not available.
+
+        Args:
+            compiler: object with a ``name`` and ``version`` attribute for
+                the compiler being used.
+
+        Raises:
+            UnsupportedMicroArchitecture: if the compiler version passed as
+                argument doesn't support the current micro-architecture.
+        """
+        # No information available on compiler: return None
+        if compiler.name not in self.micro_architecture.compilers:
+            return None
```

Comment:
The interface is like that as it takes into account three cases:

1. There's no information on the compiler for the current architecture
2. We know that the compiler does not support the current architecture
3. We know that the compiler supports the current architecture

Case 3. just returns the correct name. Case 1. returns `None` so that no optimization flags are added to the compilation line. Case 2. raises as we can't optimize for the architecture.

This leverages the fact that exceptions in Spack are usually caught in the outermost layer and displayed to the users as meaningful error messages. If we raise on case 1. we'll need to catch  the corresponding exception to avoid it going through the stack - more code on the caller side. If we return `None` for both 1. and 2. we can't distinguish the two cases anymore.